packages/hardhat-core/src/internal/hardhat-network/provider/modules/eth.ts

```diff
@@ -500,16 +509,33 @@ export class EthModule {
 
   // eth_getBlockTransactionCountByNumber
 
-  private _getBlockTransactionCountByNumberParams(params: any[]): [BN] {
-    return validateParams(params, rpcQuantity);
+  private _getBlockTransactionCountByNumberParams(params: any[]): [BlockTag] {
+    return validateParams(params, blockTagType);
   }
 
   private async _getBlockTransactionCountByNumberAction(
-    blockNumber: BN
+    blockTag: BlockTag
   ): Promise<string | null> {
-    const block = await this._node.getBlockByNumber(blockNumber);
-    if (block === undefined) {
-      return null;
+    let blockNumber: BN | "pending";
+    let block: Block | undefined;
+
+    try {
+      blockNumber = await this._blockTagToBlockNumber(blockTag);
+    } catch (error) {
+      if (error.message.includes("Received invalid block number")) {
+        return null;
+      }
+
+      throw error;
+    }
+
+    if (blockNumber === "pending") {
+      block = await this._node.getPendingBlock();
+    } else {
+      block = await this._node.getBlockByNumber(blockNumber);
+      if (block === undefined) {
+        return null;
+      }
     }
```

Comment:
This is duplicated with `_getTransactionByBlockNumberAndIndexAction` method. I would also extract lines 522-530 to a separate method themselves.